### ISSUE TITLE
Introduce a permission to destroy only snapshots, not datasets

### ISSUE THREAD

**Issue by emtiu** [openzfs/zfs#11524]
Currently, destroying datasets and destroying snapshots are both covered by the `destroy` permission.

As per [this blogpost](https://utcc.utoronto.ca/~cks/space/blog/solaris/ZFSNoSnapshotDeleteDelegation?showcomments#comments) and [this old bug report](https://www.illumos.org/issues/5989), it would be useful to be able to delegate the _permission to destroy snapshots, but not datasets_. This permissions might be called `destroysnaps`.

There's at least two use cases:

1. Consider an environment in which users might want to take snapshots, but not create or destroy any datsets (such as: there's one dataset per home directory for all users). Here, we might want to allow users to maintain snapshots on their own (in order to save space, or maintain a tidy set of snapshots). This is currently not possible without also granting the permission to destroy datasets, which is undesirable.
2. `destroysnaps` might bring more peace of mind to anxious users (like myself): We could `allow` ourselves to destroy snapshots as a normal user, but leave the destroying of datasets to root, avoiding the horror of being one typo away from accidentally destroying a dataset.

**Comment by crabique** [comment]
Agree, every time I'm recursively destroying snapshots I'm anxious if I forgot the @ somewhere, also could be useful for enforcing capabilities in multi-tenant filesystem scenarios.

**Comment by putnam** [comment]
This seems like a no-brainer and I'd really like to see this. Besides the risky command line stuff, there are lots of snapshot management tools where you really want to give them the least amount of permissions possible. There is a *big* difference between destroying a snap and a dataset.

**Comment by allanjude** [comment]
What about bookmarks? Is that implied by the snapshot permission? Or does it need its own?

**Comment by emtiu** [comment]
> What about bookmarks? Is that implied by the snapshot permission? Or does it need its own?

Since users here are mainly concerned with the safety of preventing accidental _datset_ destruction, I'd say bookmarks could be included under `destroysnaps`.

**Comment by parke** [comment]
Fwiw, the idea of adding granular permission for snapshot destruction was mentioned in @allanjude's talk at the [OpenZFS Developer Summit 2022](https://openzfs.org/wiki/OpenZFS_Developer_Summit_2022).

*  Talk summary: [Enabling Storage Multi-Tenancy With ZFS For Containers](https://openzfs.org/wiki/OpenZFS_Developer_Summit_2022_Talks#Enabling_Storage_Multi-Tenancy_With_ZFS_For_Containers_(Allan_Jude))
*  Slide number 11: [Future Ideas](https://docs.google.com/presentation/d/1fZEQNoJJhz6pW3M8S6cg-XKrBmHSDxyc/edit#slide=id.p11) - "Allow destroy snapshots only, not all children"
*  YouTube [video](https://www.youtube.com/watch?v=9o61ozATAoQ) of the talk

**Comment by skirmess** [comment]
> What about bookmarks? Is that implied by the snapshot permission? Or does it need its own?

They need their own as destroying bookmarks can destroy your backup procedure. You don't want users, or accidents, to prevent your backup from working.

**Comment by emtiu** [comment]
> They need their own as destroying bookmarks can destroy your backup procedure. You don't want users, or accidents, to prevent your backup from working.

Good point. So `destroysnaps` and `destroybookmarks` should be separate.

**Comment by f0x52** [comment]
Has there been any progress on this, or are there particular reasons this seems to be stuck?

**Comment by emtiu** [comment]
My best guess is that no developers capable of implementing this feature are interested in it.

**Comment by joshenders** [comment]
The best way to get attention for this feature is probably to add it to the agenda and attend a monthly OpenZFS Leadership meeting: https://openzfs.org/wiki/OpenZFS_Leadership_Meeting

**Comment by bghira** [comment]
the best way to get this feature rolling is to open a half-assed pull request and hope that others come along to point out the flaws and help fix/improve it for inclusion; i do this all the time to good effect.
